### PR TITLE
fix(runtime): support Linux CUDA on driver-only hosts

### LINKED ISSUE #1778: Linux CUDA runtime fails on driver-only hosts without toolkit libraries
## Problem

The Linux installer now selects a versioned CUDA runtime from the NVIDIA driver's advertised CUDA level, but the downloaded runtime still links toolkit libraries such as `libcudart.so.13` and `libcublas.so.13` without bundling them.

On a supported Linux host with an NVIDIA driver but no CUDA toolkit, installation therefore gets past the 404 fixed by #1267 and then `serve` fails in the dynamic loader. The reporter in #1220 confirmed that adding CUDA wheel library directories to `LD_LIBRARY_PATH` works around the failure. Windows runtime packages already ship their corresponding CUDA DLL dependencies.

This is the remaining runtime-linkage half extracted from #1220. It is separate from selecting the correct `-cuda-12` or `-cuda-13` asset.

## Acceptance criteria

- Define and enforce one Linux CUDA runtime contract for driver-only hosts:
  - bundle and load the redistributable CUDA runtime dependencies from the native runtime package; or
  - reject the configuration before runtime download/startup with a precise list of required host toolkit libraries and installation guidance.
- Validate on a clean supported Linux NVIDIA host with the driver installed and no system CUDA toolkit.
- Keep CUDA 12 and CUDA 13 lanes covered.
- Add packaging or resolver regression coverage so a runtime cannot silently claim compatibility while required shared libraries are unavailable.

## Evidence

- Original report and workaround: #1220
- Installer asset-selection fix: #1267
- Related driver-versus-toolkit selection background: #1195

### PR DESCRIPTION
**Pull request by i386**
Linux users can now install and run a compatible CUDA build on a machine that has an NVIDIA driver but no CUDA toolkit. The installer falls back to the driver-supported major, while Linux CUDA runtime archives carry the reviewed toolkit dependency closure needed by the native runtime.

The packager recursively resolves same-architecture ELF dependencies, admits only the CUDA-major-matched cudart, cuBLAS, cuBLASLt, and nvJitLink families, rejects toolkit stubs and incomplete or conflicting graphs, and keeps redistributed NVIDIA objects byte-for-byte unchanged. Their distribution terms are included in the archive. Runtime manifests preserve dependency-first load order, the verifier enforces the closed graph, and resolver compatibility only treats a package as self-contained when all required libraries match the requested CUDA major.

The Linux CUDA product smoke no longer installs cudart or cuBLAS on the runner and verifies the composed runtime with `LD_LIBRARY_PATH` unset before exercising inference. It records CUDA visibility variables, host driver-library resolution and NVIDIA device nodes, then runs a shared-cudart packaged benchmark probe with inherited and strict packaged-library resolution. Any device-probe mismatch fails the lane. The suite forces runtime resolution through the staged bundle manifest so a same-version published catalog cannot describe a different artifact than the product under test.

Validation:

- `just ci-validate`
- `python3 -m unittest discover -s scripts/tests -p 'test_*.py'` — 1,098 passed, 9 skipped at exact head `a04b03638`
- `cargo test -p mesh-llm-native-runtime --all-targets` — 37 passed
- `cargo check -p mesh-llm-native-runtime -p mesh-llm`
- `cargo clippy -p mesh-llm-native-runtime -p mesh-llm --all-targets -- -D warnings`
- `just ci-shellcheck scripts/lib/lld.sh install.sh scripts/package-native-runtime.sh scripts/verify-native-runtime-package.sh scripts/ci-product-integration-smoke.sh`
- `just check-release`

Candidate-branch validation at exact head `a04b03638`:

- The current `bugfix/0.76.1` base through #1754 merged cleanly with no overlap in the CUDA diff. Ordinary PR checks are running on this exact head.
- [CUDA 12 Linux packaging and driver-only product run at `1adfb0c84`](https://github.com/Mesh-LLM/mesh-llm/actions/runs/34573044143) — artifact verification and four driver-only serving phases passed: dense standalone, OpenAI SDK, constrained restart, and dense split. The recurrent split phase then failed its prefix-cache reuse assertion identically on two attempts; that behavior is outside this packaging diff.
- [CUDA 13 device discriminator at `e9066b2a7`](https://github.com/Mesh-LLM/mesh-llm/actions/runs/34608697154) — both the inherited-environment and strict packaged-library probes reported one CUDA device on the RTX 5090, then the driver-only package served through the same four phases. It failed only at the existing recurrent split prefix-reuse assertion.
- [CUDA 13 qualification at `ac91dd963`](https://github.com/Mesh-LLM/mesh-llm/actions/runs/34618732322) — runtime and product packaging passed, including the complete CUDA 13 closure. Both probes saw one GPU, but host startup evaluated the published 0.76.0 catalog entry for the duplicate runtime identity; that older entry lacks this PR's bundled-library metadata and selected CPU. The smoke now isolates staged-bundle evaluation.
- Independent mesh2 aarch64 validation hid the CUDA 13.2 toolkit in a private mount namespace. With the staged bundle manifest evaluated, the package resolved cudart/cuBLAS from `$ORIGIN`, found one CUDA device, loaded all 31 layers on CUDA0, became ready in 14 seconds, and returned a chat completion. The host toolkit and linker state were restored unchanged.
- [Exact-head CUDA 13 qualification at `a04b03638`](https://github.com/Mesh-LLM/mesh-llm/actions/runs/34624861726) — running with shared-cudart probe enforcement and staged-bundle catalog isolation.

Fixes #1778

## Summary by CodeRabbit

- **New Features**
  - Linux CUDA runtimes now include required redistributable CUDA libraries and license information.
  - CUDA runtimes can load on systems with an NVIDIA driver but without a locally installed CUDA toolkit.
  - Runtime packages now verify architecture, dependency completeness, library ordering, and supported CUDA components.

- **Bug Fixes**
  - Improved CUDA version detection when toolkit libraries are unavailable.
  - CUDA smoke tests now validate packaged runtime contents without runner-installed CUDA libraries.

### CODE REVIEW

File: scripts/linux-native-runtime-deps.py [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+) -> pathlib.Path:
+    _check_arch(source, arch, "search dependency")
+    destination = lib_dir / dependency
+    if destination.exists():
+        if _sha256(destination) != _sha256(source.path):
+            raise RuntimeError(
+                f"conflicting packaged dependency {dependency}: "
+                f"{destination} and {source.path}"
+            )
+        return destination
```
[the remark this answers is not in the document]
Comment (i386) [reply scripts/linux-native-runtime-deps.py:348]: Fixed in f1edcfb92. `_copy_dependency` now rejects existing symlink destinations before hashing, and `collect_dependencies` fails explicitly when an iteration makes no progress. Added regressions for both paths; the full 1,085-test CI validation suite passes locally.

File: scripts/package-native-runtime.sh [hunk 2]
```diff
@@ -432,6 +433,125 @@ collect_runtime_libraries() {
...
+    mkdir -p "$stage_dir/licenses"
+    cp "$source" "$stage_dir/licenses/NVIDIA-CUDA-LICENSE.txt"
+    license_paths+=("licenses/NVIDIA-CUDA-LICENSE.txt")
+}
+
+collect_linux_cuda_dependencies() {
+    case "$TARGET_TRIPLE/$BACKEND" in
+        *linux/cuda|*linux/cuda-blackwell) ;;
+        *) return 0 ;;
+    esac
```
[the remark this answers is not in the document]
Comment (i386) [reply scripts/package-native-runtime.sh:544]: Fixed in f1edcfb92. The matcher is now `*linux*/cuda|*linux*/cuda-blackwell`, and a portable shell harness proves that `x86_64-unknown-linux-gnu/cuda` invokes both collector operations. I also corrected the Linux ELF fixture search path that caused the prior CI failure; 1,085 script tests now pass locally.

**Review by ndizazzo — COMMENTED** [review]
Independent review at head ac91dd963, plus a driver-only validation on a Jetson AGX Orin (mesh2: aarch64, driver 595.78, toolkit 13.2 from debs).

## Validation

Method: built the PR runtime package (aarch64 cuda13, sm_87) and host binary, then ran inside a private mount namespace with the CUDA 13.2 lib dir masked and a driver-only `ld.so.cache` bound over `/etc/ld.so.cache`. Inside: `ldconfig -p` has 0 cudart/cublas entries, the system `libcudart.so.13` will not `dlopen`, and the pre-PR 0.72.1 runtime fails with `libcudart.so.13 => not found`. The host outside the namespace is untouched.

Result with the bundle manifest evaluated: the cuda13 bundle is accepted, `ggml_cuda_init: found 1 CUDA devices` through the packaged `libcudart.so.13` (`ldd` resolves cudart/cublas/cublasLt to `$ORIGIN`, `libcuda.so.1` to the driver dir), all 31 layers on CUDA0, ready in 14 s, chat completion returns tokens. Same on the normal host. The packaging change does what it claims.

## Findings

1. **The new smoke guard cannot fail, and its probe does not exercise the bundled cudart.** `scripts/ci-product-integration-smoke.sh` `verify_artifact_local_cuda_runtime`: every probe classification, including "packaged runtime cannot expose the CUDA device", only echoes and returns 0. And `tools/mesh-llm-gpu-benchmark` is built by bare `nvcc` (`scripts/package-native-runtime.sh`, `cuda_selected_compiler` call), whose default is `-cudart static`: on the mesh2 package `ldd` lists no `libcudart` at all, so `env -u LD_LIBRARY_PATH` toggles nothing and the `ldd | grep libcudart` line always prints "No CUDA libraries". If packaging regresses to host-linked cudart, this lane stays green. Suggest `-cudart shared` for the tool and `return 1` on the mixed classifications. Small, worth doing in this PR since the guard is the point of the smoke change.

2. **Bundled artifact accepted when the driver major is unknown** (`resolver.rs` `evaluate_cuda_toolkit_major`). With `driver_max_major == None` (no `nvidia-smi`, profile kept alive by `gpu_arches`) and no toolkit, the old path rejected; the bundled early return now accepts a cuda13 package on a 12.x driver and it fails at load with "insufficient driver". Narrow, but consider requiring `driver_max_major.is_some()` for the early return, or documenting the assumption.

3. **Catalog entry wins over an identical-identity bundle** (pre-existing, `crates/mesh-llm-runtime-install/src/manifest.rs` `append_bundle_artifacts`). A release-profile host stamped 0.76.0 consults the published v0.76.0 `native-runtimes.json`; the bundle has the same id/mesh_version/skippy_abi, so the catalog entry is what the resolver evaluates while the bundle dir is what gets served. The published entry predates this PR and lists no bundled cudart, so on the driver-only host the resolver said "no CUDA toolkit detected", downloaded the CPU runtime, and the model open failed with `unknown selected backend device: CUDA0`. With the catalog unreachable the same bundle passed. This does not affect a fresh 0.76.1 release (the catalog is regenerated by this script) but it means any shipped bundle that differs from the published catalog for the same version is judged by the stale description. Follow-up: evaluate the served bundle's own manifest when it duplicates a catalog entry. It should land before the first bundled release goes out through `install.sh`, which now picks a CUDA archive from `nvidia-smi` alone.

4. Low: `verify-native-runtime-package.sh` now skips `verify_ldd_resolution` for every CUDA entry, including the relocatable `libggml-cuda.so` and tools, so real-loader `$ORIGIN` resolution is no longer proven for them. Restricting the skip to the unpatched NVIDIA libs keeps that check.

5. Low: `verify` now rejects pre-PR CUDA artifacts ("unpackaged Linux runtime ELF dependencies"). Fine if never run on old artifacts; noting it.

No defects found in `artifact_bundles_cuda_runtime` name matching, `collect_linux_cuda_dependencies` ordering and manifest inclusion, RUNPATH handling, or `linux-native-runtime-deps.py`.

Test gaps: package tests stub `python_bin`, so the real collect/order wiring and the apt `copyright` license fallback are untested; no resolver test for `driver_max_major: None` with a bundled artifact; no verify test that an out-of-order manifest fails; the smoke test stubs both verifier and benchmark, so item 1 is invisible to it.

Landing recommendation: fix item 1 here, track 2 to 4 as follow-ups, then squash onto `bugfix/0.76.1`.

**Comment by i386** [comment]
Addressed the blocking smoke-guard finding at exact head `a04b03638`:

- Linux CUDA benchmark builds with `-cudart shared`, so `ldd` and the strict probe exercise the packaged `libcudart`.
- Both mixed probe outcomes and total device-access failure now return nonzero before product phases start.
- The product suite points native-runtime catalog resolution at its bundle fallback, so it evaluates the manifest of the staged artifact instead of a same-version published catalog entry.
- Added behavioral coverage for all three failure classifications plus an nvcc-argument test for shared cudart.

Validation at this head: shellcheck passes and all 1,098 script tests pass (9 skipped). Exact-head CUDA 13 qualification is running: https://github.com/Mesh-LLM/mesh-llm/actions/runs/34624861726

**Review by ndizazzo — APPROVED** [review]
Re-reviewed the code at `a04b03638` (no new test cycles, just reading the delta since `ac91dd963` plus a second pass over the full diff).

The blocking item from my last review is fixed properly. The Linux benchmark now links `-cudart shared` and keeps its `$ORIGIN/../lib` RUNPATH, so the strict `env -u LD_LIBRARY_PATH` probe really does resolve the bundled cudart. All three non-passing probe classifications return 1 and write to stderr, and the new smoke test covers each of them plus the `-cudart shared` argument. Forcing `MESH_LLM_NATIVE_RUNTIME_MANIFEST_URL` at an unreachable loopback address rides the documented "catalog unavailable, bundles carry the load" path in `load_release_manifest_with_sources`, so the smoke qualifies the artifact the workflow actually built rather than whatever the published same-version catalog describes. That's the right call for this lane.

I also re-read the resolver change. `artifact_bundles_cuda_runtime` now requires the exact `lib<component>.so.<required>` (or `<component>64_<required>.dll`) names for all three of cudart, cublas, and cublasLt, and the three new tests pin down accepted, partial, and wrong-major bundles. The packaging ordering (collect deps, then rewrite RUNPATH on the relocatable set only) and the manifest's `relocatable_libraries` field line up with the verifier's relaxed origin check.

Items 2 to 4 from my earlier review stand as follow-ups, not blockers: the `driver_max_major == None` early accept, the catalog entry winning over an identical-identity bundle, and `verify_ldd_resolution` being skipped for every CUDA entry. Let's track the catalog one in particular before the first bundled release goes out through `install.sh`.

Approving.
